### PR TITLE
Implement gRPC Security BBEs

### PR DESCRIPTION
**Pull request by ldclakmal**
## Purpose
$subject was done for following BBEs:

**gRPC Security**
- Service - TLS
- Service - MTLS
- Service - Basic Auth File User Store
- Service - Basic Auth LDAP User Store
- Service - JWT Auth
- Service - OAuth2
- Client - TLS
- Client - MTLS
- Client Basic Auth
- Client Bearer Token
- Client - Self-signed JWT Auth
- Client - OAuth2 client credentials grant type
- Client - OAuth2 password
- Client - OAuth2 Refresh token

Fixes https://github.com/ballerina-platform/ballerina-standard-library/issues/1150

### CODE REVIEW

File: examples/grpc-client-bearer-token-auth/grpc_client_bearer_token_auth.metatags [hunk 1]
```diff
@@ -0,0 +1,2 @@
+description: BBE on how to secure HTTP client with Bearer token auth in Ballerina.
+keywords: ballerina, ballerina by example, bbe, http, auth, jwt auth
```
Comment (BuddhiWathsala) [inline examples/grpc-client-bearer-token-auth/grpc_client_bearer_token_auth.metatags:2]:
```suggestion
keywords: ballerina, ballerina by example, bbe, grpc, auth, jwt auth
```

File: examples/grpc-client-basic-auth/grpc_client_basic_auth.bal [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import ballerina/io;
+
+// Defines the gRPC client to call the Basic auth secured APIs.
+// The client metadata is enriched with the `Authorization: Basic <token>`
+// header by passing the `grpc:CredentialsConfig` for the `auth` configuration
```
Comment (BuddhiWathsala) [inline examples/grpc-client-basic-auth/grpc_client_basic_auth.bal:5]: API doc link is not available here which I saw in other clients.

File: examples/grpc-client-mutual-ssl/grpc_client_mutual_ssl.bal [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import ballerina/grpc;
+import ballerina/io;
+
+// An gRPC client can be configured to initiate new connections that are
```
Comment (dilanSachi) [inline examples/grpc-client-mutual-ssl/grpc_client_mutual_ssl.bal:4]:
```suggestion
// A gRPC client can be configured to initiate new connections that are
```

File: examples/grpc-client-ssl-tls/grpc_client_ssl_tls.bal [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import ballerina/io;
+
+// An gRPC client can be configured to communicate through SSL/TLS as well.
```
Comment (dilanSachi) [inline examples/grpc-client-ssl-tls/grpc_client_ssl_tls.bal:3]:
```suggestion
// A gRPC client can be configured to communicate through SSL/TLS as well.
```

File: examples/grpc-service-oauth2/grpc_service_oauth2.description [hunk 5]
```diff
@@ -0,0 +1,13 @@
+// An gRPC service/resource can be secured using OAuth2 and by enforcing
+// authorization optionally. Then, it validates the OAuth2 token sent in the
+// `Authorization` metadata against the provided configurations. This calls the
+// configured introspection endpoint to validate.<br/>
+// Ballerina uses the concept of scopes for authorization. A resource declared
+// in a service can be bound to one/more scope(s). The scope can be included
+// in the introspection response using a custom claim attribute. That custom
+// claim attribute is also can be configured as the `scopeKey`.<br/>
```
Comment (dilanSachi) [inline examples/grpc-service-oauth2/grpc_service_oauth2.description:8]:
```suggestion
// claim attribute can also be configured as the `scopeKey`.<br/>
```

File: examples/grpc-service-ssl-tls/grpc_service_ssl_tls.bal [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import ballerina/grpc;
+
+// An gRPC listener can be configured to communicate through SSL/TLS as well.
+// To secure an listener using SSL/TLS, the listener needs to be configured
```
Comment (dilanSachi) [inline examples/grpc-service-ssl-tls/grpc_service_ssl_tls.bal:4]:
```suggestion
// To secure a listener using SSL/TLS, the listener needs to be configured
```

File: examples/grpc-client-bearer-token-auth/grpc_client_bearer_token_auth.out [hunk 7]
```diff
@@ -0,0 +1,14 @@
+# Create a Ballerina package.
+# Copy the generated `grpc_secured_pb.bal` stub file to the package.
+# For example, if you create a package named `client`, copy the stub file to the `client` package.
+
+# Create a new `grpc-client-bearer-token-auth.bal` Ballerina file inside the `client` package and add the client implementation.
```
Comment (BuddhiWathsala) [inline examples/grpc-client-bearer-token-auth/grpc_client_bearer_token_auth.out:5]: Shouldn’t the bal file name should use underscore?

File: examples/grpc-service-basic-auth-file-user-store/grpc_service_basic_auth_file_user_store.description [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+// enforcing authorization. Then, it validates the Basic auth token sent as the
+// `Authorization` metadata against the provided configurations. This reads data
+// from a file, which has a TOML format. This stores the usernames, passwords
+// for authentication, and scopes for authorization.<br/>
+// Ballerina uses the concept of scopes for authorization. A resource declared
+// in a service can be bound to one/more scope(s).<br/>
+// In the authorization phase, the scopes of the service/resource are compared
+// against the scope included in the user store for at least one match between
+// the two sets.<br/>
+// `Config.toml` has defined three users - alice, ldclakmal and eve. Each user has a
```
Comment (BuddhiWathsala) [inline examples/grpc-service-basic-auth-file-user-store/grpc_service_basic_auth_file_user_store.description:11]: It's better if we can use the oxford comma here.
```suggestion
// `Config.toml` has defined three users - alice, ldclakmal, and eve. Each user has a
```

**Review by BuddhiWathsala — APPROVED** [review]

**Review by dilanSachi — APPROVED** [review]

File: examples/grpc-client-bearer-token-auth/grpc_client_bearer_token_auth.out (hunk 7 above)
Comment (ldclakmal) [reply examples/grpc-client-bearer-token-auth/grpc_client_bearer_token_auth.out:5]: Ah yes. I missed that.

Comment (ldclakmal) [inline examples/grpc-client-bearer-token-auth/grpc_client_bearer_token_auth.out:5]:
```suggestion
# Create a new `grpc_client_bearer_token_auth.bal` Ballerina file inside the `client` package and add the client implementation.
```

Comment (ldclakmal) [reply examples/grpc-client-bearer-token-auth/grpc_client_bearer_token_auth.out]: Fixed with https://github.com/ballerina-platform/ballerina-distribution/pull/2083/commits/6ae2cd68bb4b6167e4c9322a5cbf9a0bbcd77ade

**Review by BuddhiWathsala — APPROVED** [review]

**Pull request merged**